### PR TITLE
try mmap, fall back to just plain file handle

### LINKED ISSUE #1302: OSError invalid argument mmap
Tried building a package with `conda-build` version `2.0.0` as well, but am unable to due to this error.

```
Traceback (most recent call last):
  File "/opt/conda/bin/conda-build", line 6, in <module>
    sys.exit(conda_build.cli.main_build.main())
  File "/opt/conda/lib/python3.5/site-packages/conda_build/cli/main_build.py", line 234, in main
    execute(sys.argv[1:])
  File "/opt/conda/lib/python3.5/site-packages/conda_build/cli/main_build.py", line 226, in execute
    already_built=None, config=config)
  File "/opt/conda/lib/python3.5/site-packages/conda_build/api.py", line 81, in build
    need_source_download=need_source_download, config=config)
  File "/opt/conda/lib/python3.5/site-packages/conda_build/build.py", line 942, in build_tree
    config=config)
  File "/opt/conda/lib/python3.5/site-packages/conda_build/build.py", line 700, in build
    croot=config.croot)
  File "/opt/conda/lib/python3.5/site-packages/conda_build/post.py", line 370, in post_build
    fix_shebang(f, prefix=prefix, build_python=build_python, osx_is_app=osx_is_app)
  File "/opt/conda/lib/python3.5/site-packages/conda_build/post.py", line 62, in fix_shebang
    mm = mmap.mmap(fi.fileno(), 0)
OSError: [Errno 22] Invalid argument
```

xref: https://github.com/conda-forge/conda-smithy/issues/289

### PR DESCRIPTION
**Pull request by msarahan**
fixes #1302

@jakirkham please test

### CODE REVIEW

File: tests/test_api_build.py [hunk 1]
```diff
@@ -18,6 +18,7 @@
 
 from binstar_client.commands import remove, show
 from binstar_client.errors import NotFound
+from pkg_resources import parse_version
```
Comment (jakirkham) [inline tests/test_api_build.py]: Are these tests getting installed with `conda-build`? If so, we need to add `setuptools` to `run` dependencies.

Comment (msarahan) [inline tests/test_api_build.py]: Good point, but no, tests are not included with conda packages of conda-build.

**Comment by jakirkham** [comment]
Thanks @msarahan. Will take it for a spin and get back to you.

**Comment by jakirkham** [comment]
So I'm trying this, but am running into a strange problem with a build based off of the merge commit from this PR. The problem goes away with `conda-build` 2.1.4 leaving our known `mmap` issue. It looks like a `conda` issue I have seen before, but don't remember the issue number ATM. Traceback shown below.

<details>

```python
Traceback (most recent call last):
  File "/opt/conda/bin/conda-build", line 6, in <module>
    sys.exit(conda_build.cli.main_build.main())
  File "/opt/conda/lib/python3.5/site-packages/conda_build/cli/main_build.py", line 317, in main
    execute(sys.argv[1:])
  File "/opt/conda/lib/python3.5/site-packages/conda_build/cli/main_build.py", line 308, in execute
    noverify=args.no_verify)
  File "/opt/conda/lib/python3.5/site-packages/conda_build/api.py", line 127, in build
    need_source_download=need_source_download, config=config, variants=variants)
  File "/opt/conda/lib/python3.5/site-packages/conda_build/build.py", line 1342, in build_tree
    permit_unsatisfiable_variants=False)
  File "/opt/conda/lib/python3.5/site-packages/conda_build/render.py", line 380, in render_recipe
    index = get_build_index(m.config, m.config.build_subdir)
  File "/opt/conda/lib/python3.5/site-packages/conda_build/index.py", line 214, in get_build_index
    platform=subdir)
  File "/opt/conda/lib/python3.5/site-packages/conda/api.py", line 24, in get_index
    index = fetch_index(channel_urls, use_cache=use_cache, unknown=unknown)
  File "/opt/conda/lib/python3.5/site-packages/conda/fetch.py", line 300, in fetch_index
    repodatas = [(u, f.result()) for u, f in zip(urls, futures)]
  File "/opt/conda/lib/python3.5/site-packages/conda/fetch.py", line 300, in <listcomp>
    repodatas = [(u, f.result()) for u, f in zip(urls, futures)]
  File "/opt/conda/lib/python3.5/concurrent/futures/_base.py", line 398, in result
    return self.__get_result()
  File "/opt/conda/lib/python3.5/concurrent/futures/_base.py", line 357, in __get_result
    raise self._exception
  File "/opt/conda/lib/python3.5/concurrent/futures/thread.py", line 55, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/opt/conda/lib/python3.5/site-packages/conda/fetch.py", line 75, in func
    res = f(*args, **kwargs)
  File "/opt/conda/lib/python3.5/site-packages/conda/fetch.py", line 117, in fetch_repodata
    timeout=(6.1, 60))
  File "/opt/conda/lib/python3.5/site-packages/requests/sessions.py", line 501, in get
    return self.request('GET', url, **kwargs)
  File "/opt/conda/lib/python3.5/site-packages/requests/sessions.py", line 488, in request
    resp = self.send(prep, **send_kwargs)
  File "/opt/conda/lib/python3.5/site-packages/requests/sessions.py", line 641, in send
    r.content
  File "/opt/conda/lib/python3.5/site-packages/requests/models.py", line 797, in content
    self._content = bytes().join(self.iter_content(CONTENT_CHUNK_SIZE)) or bytes()
  File "/opt/conda/lib/python3.5/site-packages/requests/models.py", line 730, in generate
    chunk = self.raw.read(chunk_size)
```

</details>

**Comment by msarahan** [comment]
What version of requests is this?  I'm wondering if requests is having troubles writing something to your shared folder.

**Comment by jakirkham** [comment]
The `requests` version is `2.13.0`.

Also all of this all comes from the docker image `condaforge/linux-anvil@sha256:cabd0ceddfc67e5ae9547b4f8b7e7264f49818fed4732fe8f10939e44a317b66`. So you can pull it down and play with it. Also other dependencies are listed below.

Environment details:

<details>

```yaml
name: root
channels: !!python/tuple
- conda-forge
- defaults
dependencies:
- conda-forge::anaconda-client=1.5.1=py35_0
- conda-forge::beautifulsoup4=4.5.3=py35_0
- conda-forge::ca-certificates=2017.1.23=0
- conda-forge::certifi=2017.1.23=py35_0
- conda-forge::chardet=2.3.0=py35_0
- conda-forge::clyent=1.2.1=py35_0
- conda-forge::conda=4.2.13=py35_0
- conda-forge::conda-build=2.1.4=py35_0
- conda-forge::conda-env=2.6.0=0
- conda-forge::curl=7.52.1=0
- conda-forge::expat=2.1.0=2
- conda-forge::filelock=2.0.6=py35_0
- conda-forge::git=2.8.2=4
- conda-forge::jinja2=2.9.5=py35_0
- conda-forge::libiconv=1.14=4
- conda-forge::markupsafe=0.23=py35_1
- conda-forge::ncurses=5.9=10
- conda-forge::obvious-ci=0.6.1=py35_2
- conda-forge::openssl=1.0.2h=3
- conda-forge::patchelf=0.9=1
- conda-forge::pip=9.0.1=py35_0
- conda-forge::pkginfo=1.2.1=py35_0
- conda-forge::pycosat=0.6.1=py35_0
- conda-forge::pycrypto=2.6.1=py35_0
- conda-forge::python=3.5.3=1
- conda-forge::python-dateutil=2.6.0=py35_0
- conda-forge::pytz=2016.10=py35_0
- conda-forge::pyyaml=3.12=py35_0
- conda-forge::readline=6.2=0
- conda-forge::requests=2.13.0=py35_0
- conda-forge::ruamel_yaml=0.11.14=py35_0
- conda-forge::setuptools=33.1.0=py35_0
- conda-forge::six=1.10.0=py35_1
- conda-forge::sqlite=3.13.0=1
- conda-forge::tk=8.5.19=1
- conda-forge::wheel=0.29.0=py35_0
- conda-forge::xz=5.2.2=0
- conda-forge::yaml=0.1.6=0
- conda-forge::zlib=1.2.11=0
- conda-verify=2.0.0=py35_0
- pip:
  - ruamel-yaml==0.11.14
prefix: /opt/conda
```

</details>

**Comment by msarahan** [comment]
Sorry @jakirkham this works perfectly fine for me, running the anvil docker image under MacOS Sierra, docker 1.13.1.  Please provide more detail about:

1) what your host system is (linux vs. mac vs. windows)
2) any shared folder setup you have.  Is conda-build using a shared folder as its work folder?  Is the conda install in a shared folder?  Either of these are things that I probably would always advise against - permissions errors on shared folders and filesystem weirdness are just too common.

**Comment by jakirkham** [comment]
To spell out what was briefly referenced in the other issue, it is essential that the `conda-build` `root-dir` must live in a shared/mounted directory. I used this [branch]( https://github.com/jakirkham/zarr-feedstock/tree/cb_tst_mmap ) for testing and simply ran `./ci_support/run_docker_build.sh`. That script meets this requirement with these [lines]( https://github.com/jakirkham/zarr-feedstock/blob/432909484926adbbef366ff587b4975904c7243b/ci_support/run_docker_build.sh#L20-L21 ). Please give that a try.

**Comment by msarahan** [comment]
Sorry, I fail to see the absolute necessity of the conda-build root dir on a shared/mounted directory.  You can absolutely copy the output package from some internal non-shared root dir into a shared artifact folder after the build.  That would avoid so much complexity.  Why don't you do that?

**Comment by jakirkham** [comment]
First off, I've looked into the issue above and it appears to be a `conda`/`requests` issue (as suspected). Trying to fix it seems to uncover yet another `conda` issue with using `file://` channels. So I suspect this fix is fine, but it is just difficult to test with so many `conda` issues getting in the way. I would recommend just going ahead with it in spite of these issues.

----

In response, to the last comment.

This configuration is necessary to produce the issue that we are experiencing. There may be other ways to reproduce it (e.g. using Docker Machine instead of Docker for Mac), but this is the one I'm currently able to reproduce and share.

As far as complexity, couldn't the same question be asked of using `mmap` here? Couldn't we have read the file line-by-line to check for a shebang or other relevant data instead of searching the whole file in memory? 😉 Perhaps I'm missing something.

That said, if you are wanting to understand why the mounting code was added in the first place, you will need to ask @pelson as he added it in the early days of conda-forge. In fact, this shows up in the [first commit]( https://github.com/conda-forge/conda-smithy/blob/6563111438b3a9ef3e6ec548a0ed2ee4559113e7/templates/run_docker_build.tmpl#L35L43 ) to `conda-smithy`. This leads me to believe it was/is vital to him/his org for some reason. While I don't know why, I can probably speculate as to its value.

If one would want to leverage the feedstocks to build a lot packages locally, it would be extremely convenient to link all of the `build_artefacts` directories from each feedstock to some central location. By doing this, one could build many different feedstocks using the existing build scripts and have them all leverage the same index. In fact, this index could then be used by a larger organization on a cluster to say install local builds of these packages. In fact, if the cache is populated with relevant trusted sources in advance, one could potentially use this to build in some sort of air gapped environment. Not that I'm doing any of this. It is merely my speculation of how one might use this.

There are other benefits to having the index persist on the host machine that I know are in use. For instance, one can easily spin up a fresh docker container after the build and test the built package manually or with one's own scripts. Also smaller chains of builds (as described above), can be combined with this testing strategy. Finally one can do basically anything else that they would want to do with the index present.

Whether these use cases seem compelling or not, it seems like it would be easier to fix `conda-build` than trying to fix `conda-smithy` and nearly 2000 feedstocks. Especially while other issues like automated build systems' limitations really put the squeeze on people's already limited time and our ability to do large scale fixes of this type. As it stands, I think your proposed fix works or at least should. Unfortunately my ability to test it has been blocked by `conda` bugs.

**Comment by msarahan** [comment]
Thanks for the further info.  My suggestion does not preclude a central index on a shared folder.  I am only recommending that all temporary build folders be on local, not shared storage.

Fixing conda-build is like whack-a-mole.  These permission issues happen in so many different, strange ways that I don't think I'll ever be able to nail it down completely.  For example, on Linux, if people don't mount their volumes with -Z or -z, permissions will be really strange and almost certainly broken (been there, done that).  I'd much rather have docker be a black-box, where read-only recipes are the only thing going in, and artifacts in some shared folder are the only thing coming out.  ``conda-index`` should also be run outside of the docker container, so that it can manage cross-process locking better.

**Comment by msarahan** [comment]
Merging.  Thanks for the discussion @jakirkham.  @pelson let's talk about this more when you have time.

**Comment by jakirkham** [comment]
Did this get backported/will it be in 2.1.x?

**Comment by msarahan** [comment]
It does not look like it got backported.  I'll be doing one more 2.1.x release, I think, so I'll be sure to put it in.

**Comment by jakirkham** [comment]
Thanks Mike. Much appreciated.

Unrelated, but what is the trajectory for `conda-build` releases? Are we jumping straight to 3.x after this or is there more 2.x still?

**Comment by msarahan** [comment]
Straight to 3. 3 should hopefully be almost completely compatible, just more features. It's a major version bump because the API has changed to add the new variants feature, but all old recipes should still work.

**Pull request merged**